File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/dto/Movement.java
package uk.gov.justice.digital.hmpps.keyworker.dto;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.time.LocalTime;

@AllArgsConstructor
@NoArgsConstructor
@Data
public class Movement {
    private String offenderNo;
    private LocalDateTime createDateTime;
    private String fromAgency;
    private String toAgency;
    private String movementType;
    private String movementTypeDescription;
    private String directionCode;
    private LocalTime movementTime;
    private String movementReason;

}

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/migration/Movement.kt
package uk.gov.justice.digital.hmpps.keyworker.migration

import com.fasterxml.jackson.annotation.JsonAlias
import uk.gov.justice.digital.hmpps.keyworker.model.DeallocationReason
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime

data class Movement(
  val offenderNo: String,
  val movementDate: LocalDate,
  val movementTime: LocalTime,
  val fromAgency: String?,
  val toAgency: String?,
  val movementType: String?,
  val movementReasonCode: String?,
  val directionCode: String?,
  @field:JsonAlias("createDateTime")
  val createdAt: LocalDateTime,
) {
  val occurredAt: LocalDateTime = lazy { LocalDateTime.of(movementDate, movementTime) }.value
  val deallocationReason: DeallocationReason? =
    lazy {
      when (movementType) {
        "TRN" if (directionCode == "OUT") -> DeallocationReason.TRANSFER
        "REL" if (directionCode == "OUT") -> DeallocationReason.RELEASED
        "ADM" if (directionCode == "IN") -> DeallocationReason.TRANSFER
        else -> null
      }
    }.value
}
